File: backend/src/services/planner_service.py
from uuid import UUID
from typing import List, Dict, Set
from collections import defaultdict

from ..stores.factory import get_store
from ..stores.base import (
    StoreBase,
    CourseData,
    CourseDependencyData,
    DependencyTypeEnum,
)


class PlannerService:
    def __init__(self, store: StoreBase = None):
        self._store = store
        self._deps_by_course: Dict[UUID, List[CourseDependencyData]] = defaultdict(list)

    async def _get_store(self) -> StoreBase:
        if self._store is None:
            self._store = await get_store()
        return self._store

    async def _load_dependencies(self) -> None:
        store = await self._get_store()
        deps = await store.get_course_dependencies()
        self._deps_by_course.clear()
        for dep in deps:
            self._deps_by_course[dep.course_id].append(dep)

    async def get_all_courses(self) -> Dict[UUID, CourseData]:
        store = await self._get_store()
        return await store.get_all_courses()

    async def find_path_to_course(
        self,
        target_course_id: UUID,
        passed_ids: Set[UUID],
        current_semester: int = 1,
        max_load: float = 12.0,
    ) -> List[Dict]:
        """
        Scenario 4: Generates a plan to reach a specific course as fast as possible.
        """
        await self._load_dependencies()
        all_courses = await self.get_all_courses()

        if target_course_id not in all_courses:
            return [{"error": "Target course not found"}]

        # 1. Collect all necessary dependencies recursively
        needed_ids = set()
        to_check = [target_course_id]
        while to_check:
            curr_id = to_check.pop()
            if curr_id in passed_ids or curr_id in needed_ids:
                continue
            needed_ids.add(curr_id)
            for dep in self._deps_by_course[curr_id]:
                if dep.dependency_type == DependencyTypeEnum.prerequisite:
                    to_check.append(dep.required_course_id)

        # 2. Use a greedy approach to schedule these needed courses
        roadmap = []
        current_sem = current_semester
        courses_todo = {cid: all_courses[cid] for cid in needed_ids}
        current_passed = set(passed_ids)

        while courses_todo:
            available = []
            for cid, c in courses_todo.items():
                # Check prereqs
                can_take = True
                for dep in self._deps_by_course.get(cid, []):
                    if (
                        dep.dependency_type == DependencyTypeEnum.prerequisite
                        and dep.required_course_id not in current_passed
                    ):
                        can_take = False
                        break
                if can_take:
                    available.append(c)

            if not available:
                # Potential deadlock or missing data
                roadmap.append(
                    {
                        "semester": current_sem,
                        "error": "Cannot satisfy dependencies for remaining courses.",
                    }
                )
                break

            # Sort: give priority to courses that are prerequisites for the target or unlock more things
            # For simplicity, we just take as many as fit the load
            sem_courses = []
            sem_load = 0.0

            # Offerings check
            is_odd = current_sem % 2 != 0

            # Filter by offerings
            available_offered = []
            for c in available:
                if c.available_semesters:
                    course_is_odd = any(s % 2 != 0 for s in c.available_semesters)
                    if is_odd != course_is_odd:
                        continue
                available_offered.append(c)

            if not available_offered:
                roadmap.append(
                    {
                        "semester": current_sem,
                        "courses": [],
                        "status": "Waiting for correct semester offering",
                    }
                )
            else:
                for c in available_offered:
                    if sem_load + c.workload <= max_load:
                        sem_courses.append(c)
                        sem_load += c.workload

                for c in sem_courses:
                    current_passed.add(c.id)
                    if c.id in courses_todo:
                        del courses_todo[c.id]

                roadmap.append(
                    {
                        "semester": current_sem,
                        "courses": [
                            {"id": str(c.id), "title": c.title, "workload": c.workload}
                            for c in sem_courses
                        ],
                        "total_load": sem_load,
                    }
                )

            current_sem += 1
            if current_sem > 20:
                break

        return roadmap

File: src/Application/Planner/PlannerService.cs
using CuRoadmap.Application.Common.Interfaces;
using CuRoadmap.Application.Common.Models;

namespace CuRoadmap.Application.Planner;

public class PlannerService
{
    private readonly IApplicationDbContext _context;
    private Dictionary<Guid, List<CourseDependencyData>> _depsByCourse = new();

    public PlannerService(IApplicationDbContext context)
    {
        _context = context;
    }

    private async Task LoadDependenciesAsync(CancellationToken ct = default)
    {
        var deps = await _context.GetCourseDependenciesAsync(ct);
        _depsByCourse = deps.GroupBy(d => d.CourseId)
            .ToDictionary(g => g.Key, g => g.ToList());
    }

    public async Task<List<Dictionary<string, object>>> FindPathToCourseAsync(
        Guid targetCourseId,
        HashSet<Guid> passedIds,
        int currentSemester,
        double maxLoad,
        int? goalSemester,
        CancellationToken ct = default)
    {
        await LoadDependenciesAsync(ct);
        var allCourses = await _context.GetAllCoursesAsync(ct);

        if (!allCourses.TryGetValue(targetCourseId, out var target))
        {
            return [new Dictionary<string, object> { ["error"] = "Target course not found" }];
        }

        var neededIds = new HashSet<Guid>();
        var toCheck = new Stack<Guid>();
        toCheck.Push(targetCourseId);

        while (toCheck.Count > 0)
        {
            var currId = toCheck.Pop();
            if (passedIds.Contains(currId) || neededIds.Contains(currId)) continue;

            neededIds.Add(currId);

            if (!_depsByCourse.TryGetValue(currId, out var deps)) continue;

            var groups = new Dictionary<int, List<Guid>>();
            foreach (var dep in deps.Where(d => d.DependencyType == Domain.Enums.DependencyType.Prerequisite))
            {
                if (!groups.ContainsKey(dep.AlternativeGroup))
                    groups[dep.AlternativeGroup] = new List<Guid>();
                groups[dep.AlternativeGroup].Add(dep.RequiredCourseId);
            }

            foreach (var (groupNum, altIds) in groups)
            {
                if (groupNum == 0)
                {
                    foreach (var reqId in altIds) toCheck.Push(reqId);
                }
                else
                {
                    var picked = altIds.FirstOrDefault(passedIds.Contains);
                    if (picked == Guid.Empty)
                        picked = altIds.FirstOrDefault(neededIds.Contains);
                    if (picked == Guid.Empty && altIds.Count > 0)
                        picked = altIds[0];
                    if (picked != Guid.Empty) toCheck.Push(picked);
                }
            }
        }

        var coursesTodo = neededIds
            .Where(id => allCourses.ContainsKey(id))
            .ToDictionary(id => id, id => allCourses[id]);

        var currentPassed = new HashSet<Guid>(passedIds);
        var roadmap = new List<Dictionary<string, object>>();
        var currentSem = currentSemester;

        while (coursesTodo.Count > 0)
        {
            var available = coursesTodo
                .Where(kvp => PrereqsSatisfiedByDeps(kvp.Key, currentPassed))
                .Select(kvp => kvp.Value)
                .ToList();

            if (available.Count == 0)
            {
                roadmap.Add(new Dictionary<string, object>
                {
                    ["semester"] = currentSem,
                    ["error"] = "Cannot satisfy dependencies for remaining courses."
                });
                break;
            }

            var isOdd = currentSem % 2 != 0;
            var availableOffered = available
                .Where(c => !(goalSemester.HasValue && c.Id == targetCourseId && currentSem < goalSemester.Value))
                .Where(c => c.AvailableSemesters.Length == 0
                            || (c.AvailableSemesters.All(s => s % 2 != 0) == isOdd)
                            || c.AvailableSemesters.Contains(currentSem)
                            || c.AvailableSemesters.Any(s =>
                            {
                                var cOdd = s % 2 != 0;
                                return cOdd == isOdd;
                            }))
                .ToList();

            if (availableOffered.Count == 0)
            {
                roadmap.Add(new Dictionary<string, object>
                {
                    ["semester"] = currentSem,
                    ["course_ids"] = new List<string>(),
                    ["status"] = "Waiting for correct semester offering"
                });
            }
            else
            {
                var semCourses = new List<CourseData>();
                var semLoad = 0.0;
                foreach (var c in availableOffered)
                {
                    if (semLoad + c.Workload <= maxLoad)
                    {
                        semCourses.Add(c);
                        semLoad += c.Workload;
                    }
                }

                var semCoursesOut = new List<string>();
                foreach (var c in semCourses)
                {
                    currentPassed.Add(c.Id);
                    coursesTodo.Remove(c.Id);
                    semCoursesOut.Add(c.Id.ToString());
                }

                roadmap.Add(new Dictionary<string, object>
                {
                    ["semester"] = currentSem,
                    ["course_ids"] = semCoursesOut,
                    ["total_load"] = semLoad
                });
            }

            currentSem++;
            if (currentSem > 20) break;
        }

        return roadmap;
    }

    private bool PrereqsSatisfiedByDeps(Guid courseId, HashSet<Guid> passedIds)
    {
        if (!_depsByCourse.TryGetValue(courseId, out var deps)) return true;

        var groups = new Dictionary<int, List<Guid>>();
        foreach (var dep in deps.Where(d => d.DependencyType == Domain.Enums.DependencyType.Prerequisite))
        {
            if (!groups.ContainsKey(dep.AlternativeGroup))
                groups[dep.AlternativeGroup] = new List<Guid>();
            groups[dep.AlternativeGroup].Add(dep.RequiredCourseId);
        }

        foreach (var (groupNum, altIds) in groups)
        {
            if (groupNum == 0)
            {
                if (altIds.Any(reqId => !passedIds.Contains(reqId))) return false;
            }
            else
            {
                if (altIds.All(reqId => !passedIds.Contains(reqId))) return false;
            }
        }

        return true;
    }
}
